lib/min_heap.rb

```diff
@@ -8,64 +10,123 @@ def initialize(key, value)
 end
 
 class MinHeap
-
   def initialize
     @store = []
   end
 
   # This method adds a HeapNode instance to the heap
-  # Time Complexity: ?
-  # Space Complexity: ?
+  # Time Complexity: O(log n)
+  # Space Complexity: O(1)
   def add(key, value = key)
-    raise NotImplementedError, "Method not implemented yet..."
+    new_node = HeapNode.new(key, value)
+    @store << new_node
+
+    heap_up(@store.length - 1) unless self.empty?
   end
 
   # This method removes and returns an element from the heap
   #   maintaining the heap structure
-  # Time Complexity: ?
-  # Space Complexity: ?
-  def remove()
-    raise NotImplementedError, "Method not implemented yet..."
+  # Time Complexity: O(log n)
+  # Space Complexity: O(1)
+  def remove
+    unless self.empty?
+      swap(0, @store.length - 1)
+      removed_node = @store.pop
+
+      heap_down(0)
+
+      return removed_node.value
+    end
+  end
+
+  def remove_min
```

Comment:
I'm not sure as to the purpose to this.  Keeping the heap structure with heap_up and heap_down would mean you don't need to make a loop to find the min index.